src/DCON/Ode.jl

```diff
@@ -0,0 +1,722 @@
+        #     end
+        # end
+    else
+        for ising in 1:intr.msing
+            if intr.sing[ising].psifac > odet.psifac
+                break
+            end
+        end
+    end
+    ising = max(0, ising - 1)
+
+    # Find next singular surface
+    if false # TODO: (kin_flag)
+        # while true
+        #     ising += 1
+        #     if ising > kmsing
+        #         break
+        #     end
+        #     if psilim < kinsing[ising].psifac
+        #         break
+        #     end
+        #     odet.q = kinsing[ising].q
+        #     if mlow <= nn * odet.q && mhigh >= nn * odet.q
+        #         break
+        #     end
+        # end
+        # if ising > kmsing || singfac_min == 0
+        #     psimax = psilim * (1 - eps)
+        #     next = "finish"
+        # elseif psilim < kinsing[ising].psifac
+        #     psimax = psilim * (1 - eps)
+        #     next = "finish"
+        # else
+        #     psimax = kinsing[ising].psifac - singfac_min / abs(nn * kinsing[ising].q1)
+        #     next = "cross"
+        # end
+    else
+        while true
+            ising += 1
+            if ising > intr.msing || intr.psilim < intr.sing[min(ising, msing)].psifac
+                break
+            end
+            odet.q = intr.sing[ising].q
+            if intr.mlow <= ctrl.nn * odet.q && intr.mhigh >= ctrl.nn * odet.q
+                break
+            end
+        end
+        if ising > intr.msing || intr.psilim < intr.sing[min(ising, intr.msing)].psifac || ctrl.singfac_min == 0
+            psimax = psilim * (1 - eps)
+            odet.next = "finish"
+        else
+            psimax = intr.sing[ising].psifac - ctrl.singfac_min / abs(ctrl.nn * intr.sing[ising].q1)
+            odet.next = "cross"
+        end
+    end
+
+    # Allocate and sort solutions by increasing value of |m-ms1|
+    m = intr.mlow - 1 .+ collect(1:intr.mpert)
+    if ctrl.sort_type == "absm"
+        key = abs.(m)
+    elseif ctrl.sort_type == "sing"
+        key = m
+        if intr.msing > 0
+            key = key .- intr.sing[1].m
+        end
+        key = -abs.(key)
+    else
+        error("Cannot recognize sort_type = $ctrl.sort_type")
+    end
+    odet.index = sortperm(key, rev = true) # in original Fortran: bubble(key, index, 1, mpert)
+
+    # Initialize solutions
+    for ipert = 1:intr.mpert
+        odet.u[index[ipert], ipert, 2] = 1
+    end
+    odet.msol = intr.mpert
+    odet.neq = 4 * intr.mpert * odet.msol
+    odet.u_save .= odet.u
+    odet.psi_save = odet.psifac
+
+    # Compute conditions at next singular surface
+    q = qval(equil.psilow) # Fortran: q=sq%fs(0,4) # IS THIS RIGHT? 
+    if false #TODO: (kin_flag)
+        # if kmsing > 0
+        #     m1 = round(Int, nn * kinsing[ising].q)
+        # else
+        #     m1 = round(Int, nn * qlim) + sign(one, nn * sq.fs1[mpsi, 5])
+        # end
+    else
+        if intr.msing > 0
+            odet.m1 = round(Int, ctrl.nn * intr.sing[ising].q)
```

Comment:
You may want to confirm that round works the same way in Julia as it does in the Fortran code.